src/GraphQL/TypeApi.hs

```diff
@@ -206,13 +219,68 @@ instance forall typeName interfaces fields m.
     -- Value) pairs from runFields and build a map with them.
 
     -- might need https://hackage.haskell.org/package/linkedhashmap to
     -- keep insertion order. (TODO)
     r <- forM selectionSet $ \selection -> runFields @m @fields handler selection
-    pure $ GValue.toValue $ M.fromList r
+    pure $ GValue.toValue (GValue.mapFromList r)
+
+
+-- | Closed type family to enforce the invariant that Union types
+-- contain only Objects.
+type family RunUnionType m (a :: [Type]) :: Type where
+  RunUnionType m ((Object typeName interfaces fields):rest) = HandlerType m (Object typeName interfaces fields) :<|> RunUnionType m rest
+  RunUnionType m '[] = ()
+  RunUnionType m a = TypeLits.TypeError ('TypeLits.Text "All types in a union must be Object. Got: " 'TypeLits.:<>: 'TypeLits.ShowType a)
+
+
+-- Type class to execute union type queries.
+class RunUnion m a where
+  runUnion :: RunUnionType m a -> AST.Selection -> m GValue.Value
+
+instance forall m typeName interfaces fields rest.
+         ( RunUnion m rest
+         , MonadIO m
+         , MonadThrow m
+         , RunFields m fields
+         , KnownSymbol typeName
+         ) => RunUnion m ((Object typeName interfaces fields):rest) where
+  runUnion (lh :<|> rh) fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType queryTypeName) [] subSelection))
+    | typeName == queryTypeName = buildResolver @m @(Object typeName interfaces fields) lh subSelection
+    | otherwise = runUnion @m @rest rh fragment
+    where typeName = toS (symbolVal (Proxy :: Proxy typeName))
+  runUnion _ _ =
+    queryError "Non-InlineFragment used for a union type query."
+
+instance forall m. MonadThrow m => RunUnion m '[] where
+  runUnion _ selection = queryError ("Union type could not be resolved:" <> (show selection))
+
+instance forall m ks ru.
+         ( MonadThrow m
+         , MonadIO m
+         , RunUnion m ru
+         ) => HasGraph m (Union ks ru) where
+  type HandlerType m (Union ks ru) = RunUnionType m ru
+  -- TODO we need to give good error feedback when people try to build
+  -- Union types with non-Object Types in the list. I suspect that's
+  -- hard because overlapping instances are restrictive.
+  -- we could maybe use closed type families for
+
+  -- TODO: check sanity of query before executing it. E.g. we can't
+  -- have the same field name in two different fragment branches
+  -- (needs to take aliases into account).
+
+  -- query "{ ... on Human { name } }"
+  -- [SelectionInlineFragment (InlineFragment (NamedType "Human") [] [SelectionField (Field "" "name" [] [] [])])]
+  buildResolver handler selection = do
+    -- GraphQL invariant is that all items in a Union must be objects
+    -- which means 1) they have fields 2) They are ValueMap
+    values <- map GValue.unionMap (traverse (runUnion @m @ru handler) selection)
+    case values of
+      Left _ -> implementationError "It looks like you used a non-Object type in a Union"
+      Right ok -> pure ok
```

Comment:
I've only skimmed over this whole diff hunk. I don't really understand it. I'll try to look at it later (post-merge) and come up with helpful comments / refactorings.